internal/controller/accessrequest_controller.go

```diff
@@ -159,76 +154,246 @@ func buildResult(status api.Status, ar *api.AccessRequest) ctrl.Result {
+// using the DefaultRetry backoff which has the following configs:
+//
+//	Steps: 5, Duration: 10 milliseconds, Factor: 1.0, Jitter: 0.1
 func (r *AccessRequestReconciler) updateStatusWithRetry(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		err := r.Get(ctx, client.ObjectKeyFromObject(ar), ar)
 		if err != nil {
 			return err
 		}
-		// if it is already updated skip
-		if ar.Status.RequestState == status {
-			return nil
+		return r.updateStatus(ctx, ar, status, details)
+	})
+}
+
+// updateStatus will update the given AccessRequest status field with the
+// given status and details.
+func (r *AccessRequestReconciler) updateStatus(ctx context.Context, ar *api.AccessRequest, status api.Status, details string) error {
+	// if it is already updated skip
+	if ar.Status.RequestState == status {
+		return nil
+	}
+	ar.UpdateStatus(status, details)
+	return r.Status().Update(ctx, ar)
+}
+
+// removeArgoCDAccess will remove the subjects in the given AccessRequest from
+// the given ar.TargetRoleName from the Argo CD project referenced in the
+// ar.Spec.AppProject. The AppProject update will be executed via a patch with
+// optimistic lock enabled. It will retry in case of AppProject conflict is
+// identied.
+func (r *AccessRequestReconciler) removeArgoCDAccess(ctx context.Context, ar *api.AccessRequest) error {
+	logger := log.FromContext(ctx)
+	logger.Info("Removing Argo CD Access")
+	project := &argocd.AppProject{}
+	objKey := client.ObjectKey{
+		Namespace: ar.Spec.AppProject.Namespace,
+		Name:      ar.Spec.AppProject.Name,
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, objKey, project)
+		if err != nil {
+			e := fmt.Errorf("error getting Argo CD Project %s/%s: %w", objKey.Namespace, objKey.Name, err)
+			return client.IgnoreNotFound(e)
+		}
+		patch := client.MergeFromWithOptions(project.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+		logger.Debug("Removing subjects from role")
+		removeSubjectsFromRole(project, ar)
+
+		logger.Debug("Patching AppProject")
+		opts := []client.PatchOption{client.FieldOwner(FieldOwnerEphemeralAccess)}
+		err = r.Patch(ctx, project, patch, opts...)
+		if err != nil {
+			return fmt.Errorf("error patching Argo CD Project %s/%s: %w", objKey.Namespace, objKey.Name, err)
+		}
+		return nil
+	})
+}
+
+// removeSubjectsFromRole will iterate ovet the roles in the given project and
+// remove the subjects from the given AccessRequest from the role specified in
+// the ar.TargetRoleName.
+func removeSubjectsFromRole(project *argocd.AppProject, ar *api.AccessRequest) {
+	for idx, role := range project.Spec.Roles {
+		if role.Name == ar.Spec.TargetRoleName {
+			groups := []string{}
+			for _, group := range role.Groups {
+				remove := false
+				for _, subject := range ar.Spec.Subjects {
+					if group == subject.Username {
+						remove = true
+					}
+				}
```

Comment:
Add `break` for early exit on match
```suggestion
				for _, subject := range ar.Spec.Subjects {
					if group == subject.Username {
						remove = true
						break
					}
				}
```

Reply:
done